[MRG] ENH Disassemble check estimator (#14381)
scikit-learn/scikit-learn: doc/developers/contributing.rst

selection tools such as :class:`model_selection.GridSearchCV` and
:class:`pipeline.Pipeline`.

> Setting `generate_only=True` allows for checks to be disassembled when

Reviewer: I feel like *disassembled* is a technical term that most won't understand (I didn't).

Maybe use the same kind of sentence as in the docstring

testing an estimator. This feature can be used to parameterize tests in pytest
as follows::

  from itertools import chain
  import pytest
  from sklearn.utils.estimator_checks import check_estimator
  from sklearn.utils.estimator_checks import set_check_estimator_ids
  from sklearn.svm import LinearSVC
  from sklearn.linear_model import LogisticRegression

  estimators = [LinearSVC, LogisticRegression]

  @pytest.mark.parametrize(
      'estimator, check',

>       chain.from_iterable(check_estimator(est, generate_only=True)

Reviewer: Can `check_estimator` just return an iterator? It's a bit inconvenient that we have to chain manually

Author: This would work if `check_estimator` can be passed multiple estimators, and then generated everything on the spot.

Reviewer 2: Well or we can make this example an example for just one estimator and it would be easier ;)
Or do both? First one estimator and then two, so we have the simple case first?

                          for est in estimators),
      ids=set_check_estimator_ids)
  def test_sklearn_compatible_estimator(estimator, check):

>       check(estimator)

Reviewer: Currently this example is strictly equivalent to just calling `check_estimator` right? It would be useful to illustrate how to filter-out some tests (if that's the ultimate goal?)

Reviewer 2: No, it's radically different and that should clearly be better documented.
if you call ``check_estimator``, it will fail on the first failing test and not run any other tests. This here will run all tests and report which ones are failing.

Reviewer 2: Looks good to me apart from this. This difference should be explained and highlighted more!

selection tools such as :class:`model_selection.GridSearchCV` and
:class:`pipeline.Pipeline`.

Setting `generate_only=True` returns a generator that yields (estimator, check)
tuples where the check can be called independently from each other, i.e.

> `check(estimator)`. This allows all checks can run independently and report the

Reviewer 2: ```suggestion
`check(estimator)`. This allows all checks to be run independently and report the
```

selection tools such as :class:`model_selection.GridSearchCV` and
:class:`pipeline.Pipeline`.

Setting `generate_only=True` returns a generator that yields (estimator, check)
tuples where the check can be called independently from each other, i.e.
`check(estimator)`. This allows all checks to be run independently and report
the checks that are failing. In pytest, we can parameterize the check as
follows::

  import pytest
  from sklearn.utils.estimator_checks import check_estimator
  from sklearn.utils.estimator_checks import set_check_estimator_ids
  from sklearn.linear_model import LogisticRegression

  @pytest.mark.parametrize(
      'estimator, check',

>       check_estimator(LogisticRegression, generate_only=True)

Reviewer: ```suggestion
      check_estimator(LogisticRegression, generate_only=True),
```

syntax error ;)

      ids=set_check_estimator_ids)
  def test_sklearn_compatible_estimator(estimator, check):
      check(estimator)

> When testing more than one estimator, python provides a `chain.from_iterable`

Reviewer: I'm not super convinced of the benefit of this example. This is basically advertising for `chain`.

I'd rather have an example explaining how using `-k` on the ids can help selecting a specific check. Also, the `set_check_estimator_ids` should be briefly explained.

to run checks for many estimators::

>   import chain

Reviewer: ```suggestion
  from itertools import chain
```

  from sklearn.svm import LinearSVC

  estimators = [LogisticRegression, LinearSVC]

  @pytest.mark.parametrize(
      'estimator, check',
      chain.from_iterable(check_estimator(est, generate_only=True)

>                           for est in estimators)

Reviewer: ```suggestion
                          for est in estimators),
```

selection tools such as :class:`model_selection.GridSearchCV` and
:class:`pipeline.Pipeline`.

Setting `generate_only=True` returns a generator that yields (estimator, check)
tuples where the check can be called independently from each other, i.e.
`check(estimator)`. This allows all checks to be run independently and report
the checks that are failing. In pytest, checks can be run on multiple

> estimators by using `chain.from_iterable`::

Reviewer: Sorry if that wasn't clear before but I meant that the example with `chain.from_iterable` isn't super useful IMO. And as Andy previously said we should definitly start with the simpler example that only uses 1 estimator.
